### ISSUE TITLE
Nothing happens when trying to import

### ISSUE THREAD

**Issue by johnstonesnow** [thunderbird/import-export-tools-ng#543]
I have a load of local folders in Apple Mail (NOTE: It's running on OSX 10.14 Mojave, old style)i. exported these from Mac Mail which took a few hours at least. 

I have followed instructions to import into TB via this plugin. Tried every option possible but fails every time with message: "Import Done 0/0"

Any ideas appreciated.

**Comment by cleidigh** [comment]
@johnstonesnow 
What explicitly are you importing, a single mbox or a structure produced from the script we discussed?
@cleidigh

**Comment by johnstonesnow** [comment]
I exported all my local mail using Mac Mail's export mailboxes function. So I have a nested tree structure with folders and mboxes. 
I haven't (couldn't!) use any 'scripts'.
I read somewhere that because I am using a very old OS (Mojave), the Mail export should work with this plugin.

**Comment by cleidigh** [comment]
@johnstonesnow 
As far as I have seen, nothing is different with Mojave.
For apple structured exports you have to transform the structure to Thunderbird structure with the script. IETNG can then import.
@cleidigh

**Comment by johnstonesnow** [comment]
Ok, that would explain it. But I have no idea what script, I know the thread you're referring to but I can't follow it now it's gone into technical stuff. I am not a dev, but even if I was, I thought the last I read was that the script is broken and nobody has time to fix it, is that right or is there something working that you think it might be worth trying to learn how to run?
Thanks

**Comment by cleidigh** [comment]
@johnstonesnow 
No problem, I will back up.
Here is the script and instructions 
https://github.com/thunderbird/import-export-tools-ng/wiki

The most important thing is to make a backup of your export because the script modifies the tree in place. Since it might take a couple of tries you want the backup to start over.
We thought the script might have a problem, but that was a user issue. I did a small transform on Windows.
@cleidigh

**Comment by cleidigh** [comment]
@johnstonesnow 
You will want your backup on your system ready to recopy.
First thing you want to do is run the script with no arguments just to see python and the script are good.. Next try using the path to the top of the export. That directory should have a subdirectory with a .mbox suffix.
Report the resulting structure.
Christopher

**Comment by cleidigh** [comment]
download the script with the raw link.
Once you have python 3, you should be able to run it.
Christopher

**Comment by johnstonesnow** [comment]
> @johnstonesnow You will want your backup on your system ready to recopy. First thing you want to do is run the script with no arguments just to see python and the script are good.. Next try using the path to the top of the export. That directory should have a subdirectory with a .mbox suffix. Report the resulting structure. Christopher

Thanks, but I hardly understood a word. I have backup (duplicate folder) of my exported tree.
No idea how to run a script, don't know what arguments are, sorry!

**Comment by johnstonesnow** [comment]
I clicked the "raw" button and got a full page of the code/script

**Comment by johnstonesnow** [comment]
do I copy that and paste into terminal or something?

**Comment by johnstonesnow** [comment]
Hmm. I installed the python3 installer package. it completed successfully. To confirm I went to terminal and typed "python" again, hoping to see V3 installed but it still says V2:

"Python 2.7.10 (default, Feb 22 2019, 21:55:15) 
[GCC 4.2.1 Compatible Apple LLVM 10.0.1 (clang-1001.0.37.14)] on darwin
Type "help", "copyright", "credits" or "license" for more information."

Do I need to reboot for it to take effect?

**Comment by johnstonesnow** [comment]
I have to run out but will be back in an hour or two to make some attempt at running this script. Just bear in mind I have no clue about making directories in terminal, or even finding a directory. Only way I have ever got a file location into terminal is by dragging and dropping a finder folder at the end of a command. 
I will have a go later on but any dumba55-friendly tips would be appreciated :)

**Comment by cleidigh** [comment]
yes
copy and save as the script name

hmm I don't know how to make python 3 the default.
Did the installer mention anything on this?
I don't want to mess up your current system especially since you are not a developer.
The other possible approach is to use your Linux target which probably has python 3
Christopher

**Comment by cleidigh** [comment]
just looked.
you should try python3
that should invoke version 3

**Comment by cleidigh** [comment]
@johnstonesnow 
Ok, we will do our best.
I wish I had a solution that was simpler, but it's the nature of a lot migration scenarios.
I think we can do it, I will do my best. Once you try the python3 command we will go from there.
Not everyone is a developer, no dumba55 required %-)
Christopher

**Comment by johnstonesnow** [comment]
hehe, thanks. 
Re messing up this machine, no worries there. It's my last iMac (a very old one that's being used for zeroing my rifle once this job is done, and no I am not joking). 

python3 - that did it :) ...

"Python 3.12.2 (v3.12.2:6abddd9f6a, Feb  6 2024, 17:02:06) [Clang 13.0.0 (clang-1300.0.29.30)] on darwin
Type "help", "copyright", "credits" or "license" for more information."

So... "copy and save as the script name". 
By that do you mean use the browser menu to "save as html" to save to my computer then rename the file "apple2mboxsbd.py" ?

**Comment by cleidigh** [comment]
@johnstonesnow 
I'm on East coast time, but usually online by 12:00. I will be online until around 6:30. Let me know when you want to "live walk through"
In the meantime I will post some steps and notes. I will also fire up my MacAir which I think has Mojave also.
Christopher

**Comment by johnstonesnow** [comment]
Hi, that's extremely kind of you. I am sitting down for an hour (minimum) now. 
I am hopeful it won't take too much time, just need to learn how to run the script. I do have Python3 installed now, so that's a step in the right direction. I have the Mac Mail export (with duplicates) ready to try a conversion. 
thanks very much

**Comment by cleidigh** [comment]
@johnstonesnow 
Sorry, Just got moved to bed, I'm disabled with ALS so do programming by eyes and don't have a lot of time now. If you are still on we can do a bit. I would like you to create a new directory right under your home directory for ease. Name it mboxcvt. We will download the script there next. We can do these first few steps with Finder then we will have to use the terminal.
C

**Comment by cleidigh** [comment]
Ha, you are supposed to be John (Snow of Game of Thrones?) %-)
I will give some more steps, but will need some interaction after. I also have to be brif as typing by eyes is onerous.

Next step, download script:
- Click script link and look for download button on top right,, box with down arrow in it
- Should download to your download dir, then move to our mboxcvt dir
- Now open a terminal 
- Change to the home dir with 
cd ~
- change to mboxcvt
cd mboxcvt
- See if the script runs
python3 applembox2sbd.py
- you should see it gripe about not having a IN argument (parameter to cmds)

Setting up other directories, you can do regardless of above working 
- Under mboxcvt make two directories exptree and exporig
- copy your export tree into both directories 
- Explain the the top levels of directories in exptree

We will have to interact tomorrow any time after 12 Eastern time.
BTW I always love to know where my users are if you don't mind saying...
Christopher

**Comment by johnstonesnow** [comment]
Yes, well spotted, but you're forgetting I have to do sentry duty to watch out for white walkers :-D
Unfortunately for me I'm in the UK. Typing by eyes, that must be quite tiring. Will do my best to keep this as short (your end) as possible.

OK, step by step....

**- Click script link and look for download button on top right,, box with down arrow in it**

Ok, by "script link" I assume you mean the one that goes to this page - https://github.com/thunderbird/import-export-tools-ng/blob/master/utility-scripts/apple2mboxsbd.py
Assuming so, I have clicked the download button (one with tooltip: "Download Raw File").



**- Should download to your download dir, then move to our mboxcvt dir**

I downloaded it straight into mboxcvt dir



- **Now open a terminal
- Change to the home dir with
- cd ~
- change to mboxcvt
- cd mboxcvt
- See if the script runs
- python3 applembox2sbd.py**

ha, no such file or directory. Get this, I remembered one of about 3 commands I know (maybe I was lying when I said I wasn't a dev :) !) and did a "ls" to list contents. Then realised you put the 2 in the wrong place :D - So I ran:

python3 apple2mboxsbd.py



- **you should see it gripe about not having a IN argument (parameter to cmds)**

Yep! Output:

"usage: apple2mboxsbd.py [-h] [--verbose] [--dry-run] IN
apple2mboxsbd.py: error: the following arguments are required: IN"




- **Setting up other directories, you can do regardless of above working
- Under mboxcvt make two directories exptree and exporig**

I am assuming I can do this in Finder so will do it that way. I think I remember mkdir being the command but lack of confidence leads me to Finder wherever possible.

Ok, now we're cooking on gas :)
Here's Finder now:
![Screenshot 2024-03-06 at 14 27 06](https://github.com/thunderbird/import-export-tools-ng/assets/44774538/f5aaf66e-30a0-4f9c-8681-5c4c6ca37a7f)




Ok, these last two bullets I got a bit stuck with. 

**- copy your export tree into both directories**

Assuming I know what you mean but want to be 100% sure before I do anything further. by "export tree", am i right in thinking you mean the entire folder full of exported folders from Mac Mail? If so, you mean my folder called "mailboxes export" which is on my desktop, looks like this:

![Screenshot 2024-03-06 at 14 36 50](https://github.com/thunderbird/import-export-tools-ng/assets/44774538/9fb8abef-2edb-46a2-90fe-1e06bf86ea78)

So should I select all contents OF 'mailboxes export', and copy those into the two folders we created? 
I guess there's no harm in going ahead and doing that anyway, can always delete. 
This is done, now both of those folders look like this:

![Screenshot 2024-03-06 at 14 44 47](https://github.com/thunderbird/import-export-tools-ng/assets/44774538/f7a7f303-a29d-469c-a928-5f531ee1f16c)




**- Explain the top levels of directories in exptree**

Wasn't sure but maybe you just mean for me to show what they look like, hopefully done that above. 

If my assumptions are correct, I think we are now at the point you hoped I'd be. :)

Thank you again, sorry for all the typing but seems easier for you to read (every detail of my actions) rather than me asking and waiting for you to answer, and have to type more needlessly.

**Comment by cleidigh** [comment]
@johnstonesnow 
Excellent, you were under selling yourself!
So now we have our setup ready for converting.
from terminal, get to the mboxcvt dir
Then:
python3 apple2mboxsbd.py ./extptree

Then screenshot the exptree directory, with luck we will have our transform.
Christopher

**Comment by johnstonesnow** [comment]
Finished! 
(I am sure this doesn't need saying but just to be completely thorough, i removed the extra "t" from the command above "./extptree"
It ran and "finished!"

I watched Finder live as it happened, like magic. :) 

Old on left, New on right

![Screenshot 2024-03-06 at 18 39 45](https://github.com/thunderbird/import-export-tools-ng/assets/44774538/b9064fc7-b45b-4558-a3d7-2e2a5c428fd0)


Looks like it worked to my untrained eye. 

25 items in each folder
The old folders are still folders but with .sbd appended. 
The old .mbox files have the same name but the .mbox has been removed. 

Are we looking good? 

thanks so much for this Christopher

**Comment by cleidigh** [comment]
@johnstonesnow 
Fantastic, no problem!
Now the tree can be imported. Do you want to do that on Linux?
C

**Comment by johnstonesnow** [comment]
Great!
That was going to be my next question. 
Since this is all about moving to Fedora, would it be best to run Thunderbird on the Fedora machine (much more powerful machine), move these files over with USB stick and run the import there? I can run on Mac, whichever you think is best, the only consideration I know of is the performance difference. This machine is an old dog, the Fedora one is a new top spec AMD thinkpad. 
What say you Mr C? :)

**Comment by cleidigh** [comment]
Let's go for broke Mr. Snow, warp speed ahead!
Get latest Thunderbird and ImportExportTools up on Fedora and then we can import there.
C

**Comment by johnstonesnow** [comment]
Englishman's desk sinks into the floor, new one comes down from ceiling, chandeliers barely move....

**Comment by cleidigh** [comment]
you stumped me there...

**Comment by johnstonesnow** [comment]
I was just switching workstation, maybe I overdid the description a little :P

**Comment by johnstonesnow** [comment]
Running updates on Fedora briefly, then will install TB.

**Comment by johnstonesnow** [comment]
Ok, TB installed. ImportExportTools installed

**Comment by johnstonesnow** [comment]
I guess I need to put exptree folder and contents on a stick to bring over so will do that now

**Comment by johnstonesnow** [comment]
Can I ask a question...

Hypothetically, if I imported a folder tree into thunderbird (with IETNG or without), then repeated that process, would it append a (2) onto the end of folders?

No, I haven't done any such thing! Just curious, don't want to confuse things at this point, but I THINK I have TWO exports of my mac mail, a much older one, and this recent one we converted with the script. The older one (I just realised) has 5 yrs of older emails than the newer one. 

So I am considering running the script on the older Mac Mail export tree, but the folder names are the same. So I wonder if best to append a 2 to all the folders and mbox files in the Mac tree, before conversion. Or just leave it and (as per question above) maybe TB will just append the names of the folders on import, and I can then drag and drop emails into the relevant folders, before deleting the now superfluous '2' folders. If that makes no sense whatsoever, let me know!

**Comment by cleidigh** [comment]
@johnstonesnow 
If you import two folders at the top level IETNG will add a suffix. All the folders under can have the same names.
I would create two folders under Local Folders one for new and one for old.
Then right click and use import mbox choose all mboxes with sbd structure.
Then choose the exptree directory.
You may want to copy the usb over first.
C

**Comment by johnstonesnow** [comment]
Thanks. not sure I fully understand, here's what I think I should do (hoping you can catch me if I am going wrong anywhere):

1. Create 2 Local Folders in TB. Let's call them Import1 and Import2. 

2. Right click on Import1,  then choose this option:

![Screenshot 2024-03-06 at 20 25 51](https://github.com/thunderbird/import-export-tools-ng/assets/44774538/16972c7b-3594-44f9-884e-126175d64948)

3. Point that import at the exptree folder, the one we created and which contains our nicely converted MacMail export. 

4. Run script again on the older MacMail export tree, and repeat the process for that one, except this time I'd right click on Import2 folder in TB sidebar, to import those folders into Import2, from where I can drag and drop to sort out later. 

Does that sound about right?

Thanks

**Comment by cleidigh** [comment]
Exactly

**Comment by johnstonesnow** [comment]
Ok, I have to go eat then will be back to crack on with the above. I won't expect you to be online then but will drop some updates here, just because I know how much my ramblings mean to you. :D 

In case you have any useful input (that's fairly rhetorical at this point!) - two questions:

1. Spam! 10-15 yrs on Mac Mail, I relied completely on SpamSieve (paid) which only runs on Mac, and was extremely effective (after much training). I have no idea how spam is fought off these days without it! If you have any suggestions (over and above just training TB's "mark as junk" feature, I'd be keen to hear them. 

2. Betterbird - What are your thoughts? A while back I read up a lot on it and it seemed to have many functions I'd miss or want in Thunderbird. I was planning on installing it once I have all my mail set up properly. Your views are welcome on this too, if you have any?

So grateful to you, which leaves me with my final question, please let me know how I can 'donate' to you in some way. Your help has been above and beyond anything I would ever have hoped for, never mind expected. I'd like to send a small token of my gratitude for all this support if I may.

**Comment by cleidigh** [comment]
@johnstonesnow 
As far as spam is concerned, Thunderbird has gotten better, but nothing beats Google. Training with your junk should help.

I don't have direct experience with BB, but I prefer the official release. There is a big team behind it and I personally would not support issues with it. Too much to handle.

It's always nice working with patient and appreciative people!

Donations are always very nice. Reviews are also much appreciated! You can do both from the addon home page: 
https://addons.thunderbird.net/en-US/thunderbird/addon/importexporttools-ng/?src=hp-dl-mostpopular

Cheers
Christopher

**Comment by johnstonesnow** [comment]
Thanks. 

Spam - When you say TB has gotten better, so you're talking about just using the standard install options, and train it. That works? If so I'd rather stay as 'vanilla' as possible. You know what these AddOns are like. HAHAHA, sorry, that was payback for the JohnSnow gybes. :D Google - I assume you mean gmail, I don't use gmail. No worries, sounds like I should give vanilla TB filtering a good go before looking any further. 

BB - Hmm, in that case I will stick with TB myself. I heard about BB a long while ago and I suspect it was then useful for a few things that TB was struggling with or couldn't do at the time, but I suspect that's old news now. 

I have to run now, have import2 going as we speak, Import1 looked good (haven't checked any folders but since everything is marked unread i can see the numbers of emails on each folder and certainly doesn't look like an empty import like before you came along to save the day!

I will love you and leave you (British saying, not sure if it translates over there, hope so!). THANK YOU SO MUCH. 

I will be back in a few hours to check things over, try to replicate tons of Mac Mail rules/filters into TB, see how to get contacts over (manually probably!), and somewhere in there I will certainly find time to write a review and drop some proper gratitude your way too.

You've been truly *amazing* here.

Mark

**Comment by johnstonesnow** [comment]
PS Just in case I should let you know. 
I saw the import complete and msg at bottom left corner said: "Import Done 45/65" Is that ok or should I check carefully through email folders?

**Comment by cleidigh** [comment]
Yes I would check, never know what mbox issues one might have. Could be empty folders.
Your appreciation is well noted, and thank you.
C

**Comment by johnstonesnow** [comment]
Do the 45/65 numbers relate to folders?  I didn't think there were that many (even 45) but I will double check, I could be wrong. One thing that makes things much harder is how TB sorts all folders with newest email at bottom, instead of top like Mac! Looking for a solution to that so I can check each folder more carefully.

**Comment by johnstonesnow** [comment]
So far all folders have MORE in (just one or two, or a few in some cases). That's fine by me! Will check all

**Comment by cleidigh** [comment]
Should be folders. Check against your tree in finder.

**Comment by cleidigh** [comment]
Some messages may have been deleted but not compacted out.

**Comment by johnstonesnow** [comment]
Will do thanks. 

Just a thought, re the larger numbers, is it possible it's due to TB having every email as a separate on in the count, but Mac Mail may be counting them by 'conversation', so in a three email thread, it counts as one on mac, but 3 on TB?

**Comment by cleidigh** [comment]
Should be same Thunderbird to Thunderbird

**Comment by johnstonesnow** [comment]
Not sure I understand. I am going from Mac Mail to Thunderbird?

**Comment by cleidigh** [comment]
oh right, that might be correct

**Comment by johnstonesnow** [comment]
Ok, looks like there' may be a problem.

Example: 

One folder in particular, on Mac Mail it shows as having 37,374 messages in it.
in TB on Fedora the same folder after the import contains 5127 messages. 

Same folder on Mac contains emails back to 2006 but in TB only goes back to Feb 2023. 

Also, the emails in that folder on TB seem duped:

![Screenshot from 2024-03-06 22-52-08](https://github.com/thunderbird/import-export-tools-ng/assets/44774538/3701e09c-00c2-448e-b917-b3e946e8ddbf)

I wonder if I should start again after runing a "rebuild" of mailbox on Mac. Not sure it can handle doing so, but could try.

**Comment by johnstonesnow** [comment]
Please don't feel any pressure to respond now. This is not urgent. But I will wait for your thoughts before doing anything else in case you have ideas. Thanks

**Comment by cleidigh** [comment]
Ok, unfortunately it sounds like we may have to do some mbox debugging.
Isolate the folder that has that big discrepancy. We may have to look at it. Also I have beta for v14.0.2 which might have some relevant tweaks.

**Comment by johnstonesnow** [comment]
Is this something I can do on Fedora if I get that specific mbox folder onto it via USB stick? Or do I need to be on Mac?

"beta for v14.0.2 " - is this a reference to the script or the plugin/TB?

**Comment by cleidigh** [comment]
we can compare the usb export to the import all on fedora.
v14.0.2 is the next ImportExportTools release I am finishing up now.
It has a couple of mbox fixes. I am wondering if they may be relevant. 

Let's find the folder and deal with it individually.

**Comment by johnstonesnow** [comment]
ok I have that file/folder in Fedora file manager. I checked on Mac, the exported mbox for that folder reports as 2.15GB on Mac. Same here on Fedora, so the file got here :)

**Comment by cleidigh** [comment]
Ok now on Thunderbird, right click on the folder and choose the IETNG option open folder location. Then get the file size of the folder file with no extension.

I'm going to go offline in a bit. We may have to take a deeper dive tomorrow. Unfortunately sometimes its very hard to diagnose without seeing the files.

**Comment by johnstonesnow** [comment]
Sure, no problem at all. 

Hmm. Dangit. Tried that many times, nothing happens, FileManager doesn't even open. That could be a Fedora issue though 

PS in case it helps, I don't mind if the folders get 'un-nested', would there be any better chance of it working if i just imported one folder at a time, I can  shift them around later?

**Comment by johnstonesnow** [comment]
PPS, drop this whenever you like. Please. 
I could always delete all, start again and try the new revision of the script whenever it's ready.

**Comment by cleidigh** [comment]
well I think we look at the individual mbox from the usb and import that file.

**Comment by johnstonesnow** [comment]
I tried right clicking and choosing Properties on that folder, so I can get the location from there. Also, is the rebuild worth a shot? Assuming not.

**Comment by johnstonesnow** [comment]
> well I think we look at the individual mbox from the usb and import that file.

I see. So I could create a new top level folder in Local Folders and import just that folder, shall I do that?

(Please let me know when you go offline. I don't want my replies to make you feel you have to stay any longer than you want to!)

**Comment by cleidigh** [comment]
No there is something tripping up a bunch of messages.
I'm surprised as I have done a lot of work on this.
If you look at the end of the folder do you see messages without a subject or date?

**Comment by cleidigh** [comment]
yes import separately 
I'll let you know when I have to go
This is important to me that this is solid.

**Comment by johnstonesnow** [comment]
at the very bottom (which is oldest dated email as i re-sorted with oldest at bottom) there are 3 messages (which represent just one, twice duplicated) and THOSE THREE show no Subject, but do show a date. If that helps?

I will do a manual separate import of just this troublesome folder and let you know if i get a better number of imported emails.

**Comment by cleidigh** [comment]
ok that means email separators or From lines not being detected correctly.
Tomorrow we can try my beta.

**Comment by cleidigh** [comment]
now I have to go

**Comment by johnstonesnow** [comment]
thank you, bye for now

**Comment by johnstonesnow** [comment]
PS I have stopped doing anything this end, I wont do the manual import as I expect to be wiping clean slate for another run. Til next time!

**Comment by johnstonesnow** [comment]
PPS - Final msg of today, as i fear it may be important to note:

You know how many years (nearly 2 decades!) of emails we are stretching back with my mbox files. All of that time was spent on macs, which were cloned to next machine when i bought newer machines (only a couple in that time). I DO remember having to "rebuild mailboxes" once or twice over the past 10-15 yrs. I haven't done it for a VERY long time though, and wondering if it's just worth considering, IF my old dog of an iMac is capable of it, rebuilding all those mailboxes and exporting again before trying a new conversion. 

Just food for thought, ignore me as required! 
Cheers
M

**Comment by cleidigh** [comment]
@johnstonesnow 
I think the rebuild makes sense. Import that to a new folder so we can compare to the old.
After that we want to find small folders that also have errors.
@cleidigh

**Comment by johnstonesnow** [comment]
heading to desk now to recap on above and see what to do next

**Comment by johnstonesnow** [comment]
Ok, so a rebuild in Mac Mail, a fresh export, then a fresh conversion.

Damn it I should have done this earlier, a rebuild it could take hours on this machine. I will kick it off now and see what happens. thanks Christoper

**Comment by johnstonesnow** [comment]
I think I should probably do both a rebuild and reindex, to be as sure as possible of intergrity:

https://www.computerworld.com/article/2476217/mac-os-x-how-to-fix-apple-mail-by-rebuilding-and-reindexing-the-mailbox.html

**Comment by cleidigh** [comment]
We can do some fun bash stuff on Fedora while it churns.
I was thinking about how to analyze this last night. First we should verify the folder count and look at folders skipped, second find the smallest folders with errors. We can look on Fedora for these. Also can we rebuild individual folders?

**Comment by johnstonesnow** [comment]
You and me have very different ideas of "fun" !!

If you mean on Mac Mail, yes. In fact every guide I have seen says to click on a mailbox (which i think means mail folder) and choose "rebuild". I will be here forever doing that, so I block highlighted all my local folders (all in export tree) and asked it to rebuild all. It may have hung, it started but says "41 mailboxes selected (68586 messages)". We already know there's a lot more messages than that in those folders. Maybe I should kill it

**Comment by johnstonesnow** [comment]
I killed it. I can go one at a time yes, but thing is... Mac Mail doesn't give ANY indication of whether it's working or not so I have no clue what its doing (if anything). Even the "Activity" window never shows anything even when it is dong something! Been like it for years, normal though I think. 

Ok I am at your disposal. What do you mean by check on smaller folders, in Mac Mail, in Mac Finder (exptree), or in Linux?

**Comment by johnstonesnow** [comment]
PS that Acivity Window appears to only be for send/receive activity, I just saw it flash up that it was checking for new msgs.

**Comment by cleidigh** [comment]
well if you can traverse your max first and find a couple small folders then we can check the Thunderbird import of the same to look for a delta in messages. Once we find we can rebuild those and see what happens

**Comment by cleidigh** [comment]
we can also search in bash terminal

**Comment by johnstonesnow** [comment]
I have clicked on each and every mailbox on mac mail and asked it to Rebuild. 

No response/feedback, but at least I clicked on every single one. Maybe it happens very quickly for LOCAL messages, maybe the time some say to expect it to take is due to most people doing it on IMAP folders where it clears and re-syncs everything over the web. 

Are you asking me to look at Mac Mail, or mac Finder? Do you just mean pick out a few folders on Mac Mail which dont have a lot in, and compare those to TB folders on Fedora for discrepancies?

**Comment by johnstonesnow** [comment]
I do wonder if the best thing to do would force a reindex of Mac Mail, do you think I should? I am happy to just sit here and do exactly what you suggest, one command at a time if necessary

**Comment by cleidigh** [comment]
your last statement, look in mac mail to find small folders then we look at Thunderbird

**Comment by johnstonesnow** [comment]
ok, will do some comparisons now
thanks

**Comment by johnstonesnow** [comment]
ok. Some random picks:

1. Folder called "Legal". On Mac - 102 messages. On TB - 102 messages. 
2. Planning - 7 msgs on both
3. Security - 6 msgs on both
4. Software - 13 msgs on both

So far they all seem the same. Weird! But won't hurry, will now go look at some slightly bigger ones....

**Comment by cleidigh** [comment]
good that will be easier to compare.

**Comment by johnstonesnow** [comment]
I had a theory which now looks like a failure. I wondered if Mac was counting conversations as 1, where TB counts them as however many msgs within it. But that's not the case. If that were the case, the discrepancy between H&J folders should be 20-30 emails. but it's only 7. That's odd. Shall I manually check every one to try and find the extras on TB?

**Comment by cleidigh** [comment]
no lets wait and find the smallest mac folder that has fewer messages on Thunderbird

**Comment by johnstonesnow** [comment]
OK but i have just been manually checking each email, may as well say what happened: 

I went down from newest at top in H&J folder on both machines, cross referencing every single msg from top down. 

Same until i noticed one appearing to be missing on TB, but it was actually further down, due to having a different DATE! On Mac the message date was August 29 2023, same email in same folder on TB is dated July 17 2023. 

That's pretty weird! Let me know if you want me to stop comparing and find a folder as you said in your last message, or investigate this H&J folder further

**Comment by cleidigh** [comment]
yeah I think the folders with missing messages is best start

**Comment by cleidigh** [comment]
but note for later

**Comment by johnstonesnow** [comment]
Roger that.. Will find one

**Comment by johnstonesnow** [comment]
I quickly finished the manual check through H&J, was over half way.  
I did find one (so may have missed others) instance of TB having a carbon copy duplicate of the same email, so 2 of same where only 1 of it exists in Mac folder. That may explain the slight increased numbers on soe TB folders. 

Now to find a TB folder with fewer msgs...

**Comment by johnstonesnow** [comment]
Ok. I could describe my folder tree as being in two parts, whilst they are all in the same tree, there is a section at the top which is different. I basically have:

1. Live Folders - where I store all imprtant stuff related to certain topics/people. 
2. What I call (and think of as) Archive folders. These are all subnested under a folder called Email Archves, and they contain all my sent items from past email accounts, a trash archive (I know I know!), various other old archives I keep because I'm just weird i guess. :) 

Point being:

ALL Folders in Category 1 above seem good, at least numbers wise. There are NO TB folders with FEWER msgs than the Mac Mail counterpart. Some have a few more msgs in on TB (prob just harmless random duplicates) but none have less. 
That's Good .

Category 2 folders are, by nature, damn big. I wouldn't say it's unimportant if I lose some emails in those, but it's probably less important, whilst nice to have them all. 

So I am working up through the archive folders now. 

First one is "Trash Archive" - On Mac this shows 11797 msgs. On TB that folder shows 9672 msgs

I will keep checking but give you that bit of info in case you want me to stop here. I MAY find a smaller example but suspect they are all bigger beyond this one in archive folders.

**Comment by johnstonesnow** [comment]
ok next one is "Sent Archive" and this is bigger, with a bigger problem.

Mac - 37,374 msgs
TB - 5116 msgs

**Comment by cleidigh** [comment]
ok good info and a starting point. So how about junk?
I'm hoping that may have issues and that it might be one you could send me

**Comment by johnstonesnow** [comment]
Ha, it's the only thing I don't keep, junk/spam! 

Now then, it gets weirder. Really wondering if I should force Mac  to re-index my mail folder, then try a fresh conversion and import!  Get this....

I have a folder called "Keep Archive":

Mac - 950 msgs
TB - 1431 msgs

Explain that one!! :)

Something else odd on MAC - if i sort that folder with most recent top, the first 30-50 msgs are OLDER than some lower down. They are 2008-2009, yet further down there are emails from 2010. Interestingly, they are all a different colour, i.e. marked as spam. Should still sort by date though, but they're not. 

Re-index time perhaps?

**Comment by johnstonesnow** [comment]
is there a way to find out how many msgs are in the  Keep archive folder in exptree folder, the one I imported from? Either in file manager or terminal

**Comment by cleidigh** [comment]
we can get an idea with some bash commands.
Do you want to try that?

**Comment by johnstonesnow** [comment]
I'm game if you are :)

**Comment by cleidigh** [comment]
ok did you copy exptree to Fedora, we want to work on your fast machine

**Comment by johnstonesnow** [comment]
yes

Home/Documents/Mail Import 1/exptree

**Comment by johnstonesnow** [comment]
i can move it elsewhere if easier

**Comment by johnstonesnow** [comment]
are you opposed to the idea of me forcing a rebuild of mail index on mac while we do this? Say if not

**Comment by johnstonesnow** [comment]
can you drag and drop a finder folder into terminal like you can on mac, to cheat?!

**Comment by johnstonesnow** [comment]
i'll get there if not

**Comment by cleidigh** [comment]
don't think so, but let me know if you need guidance

**Comment by cleidigh** [comment]
ok did you get there with terminal?

**Comment by johnstonesnow** [comment]
not yet, one min

**Comment by johnstonesnow** [comment]
i have spaces in folder name Mail Import 1
how do I cd to that?

**Comment by johnstonesnow** [comment]
(My answer is to rename the folder :D )

**Comment by cleidigh** [comment]
single quotes around dir

**Comment by johnstonesnow** [comment]
ha, easy when you know how!

**Comment by johnstonesnow** [comment]
here we are

![image](https://github.com/thunderbird/import-export-tools-ng/assets/44774538/b67dac48-f899-4931-bb9e-6d3de2de6d30)

**Comment by cleidigh** [comment]
so ls there, I am assuming trash will be at top

**Comment by johnstonesnow** [comment]
no its subnested under a couple of folders. wnt me to get to that dir?

**Comment by cleidigh** [comment]
yes

**Comment by johnstonesnow** [comment]
weird, woudnt work, i guess i was trying to cd to the mbox files instead of folders. I added .sbd to the cd command and think that is how to get into these sbds
working my way to it

**Comment by johnstonesnow** [comment]
blimey its a long way down!

**Comment by cleidigh** [comment]
yes all mbox files have no extension

**Comment by johnstonesnow** [comment]
ok got a bit confused.  I am now here:

~/Documents/Mail Import 1/exptree/Archive - Local.sbd/Email Archive.sbd$

Inside which is the 'Trash Archive' FILE (i assume as its green).

**Comment by johnstonesnow** [comment]
![Screenshot from 2024-03-07 21-47-24-obfuscated](https://github.com/thunderbird/import-export-tools-ng/assets/44774538/ae0fc3ea-f675-45d4-bf2e-ecf62fa1be93)

**Comment by johnstonesnow** [comment]
PS The Mac Mail re-index is underway. 

3928 of 54,469 msgs

Time Remaning 2 hrs 23 minutes

**Comment by cleidigh** [comment]
ok we can get the msg count approximately with 
grep -c 'From ' Trash Archive'
note space after From

**Comment by johnstonesnow** [comment]
working

**Comment by johnstonesnow** [comment]
still flashing

**Comment by cleidigh** [comment]
still going?

**Comment by johnstonesnow** [comment]
yep

**Comment by cleidigh** [comment]
this is an example of the difficulty of analyzing mega files

**Comment by cleidigh** [comment]
now I am questioning what it's doing

**Comment by johnstonesnow** [comment]
still just flashing cursor :(

**Comment by cleidigh** [comment]
what happens if you type some characters then return

**Comment by johnstonesnow** [comment]
nothing if type, just angle bracket thingie goes to new line

**Comment by johnstonesnow** [comment]
heres the command i used:

grep -c 'From ' Trash Archive'

i did copy paste (which usually works) but i guess i could try typig it

**Comment by johnstonesnow** [comment]
ooh, should there be another single quote, just before Trash?
like this:

grep -c 'From ' 'Trash Archive'

**Comment by johnstonesnow** [comment]
ooh, should there be another single quote, just before Trash?
like this:

grep -c 'From ' 'Trash Archive'

**Comment by cleidigh** [comment]
there is a missing single quote in front of Trash
ctrl-c or x to get out and retype
darn

**Comment by johnstonesnow** [comment]
sorry i should have looked rather than just paste. doing that now!

**Comment by johnstonesnow** [comment]
boom

9948

**Comment by cleidigh** [comment]
sorry my fault

**Comment by johnstonesnow** [comment]
Mail programs reported:

Mac - 13925 msgs
TB - 9672 msgs

**Comment by johnstonesnow** [comment]
ran again. reporting 9948. i guess you said it would be a rough idea

**Comment by cleidigh** [comment]
quick
ok the count is any line with a From which is also including in the body

**Comment by johnstonesnow** [comment]
maybe tomorrow after the reindex of mac mail has happaned, i can re-run the script to a new exptree folder and try again. I would love to know how to properly clean out Thunderbird though before doing so. Will deleting folders in TB GUI clear everything out properly or best to do in terminal (or uninstall the flatpak of TB)?
 what do ya reckon Sire?!

**Comment by cleidigh** [comment]
having eye trouble, going to have to stop

**Comment by cleidigh** [comment]
if you drop the -c we can get output and get exactly the message count

```
grep 'From ' 'Trash Archive' | out.txt
```
Then send me out.txt
test1@kokkini.net

The delta will be From lines in body

**Comment by cleidigh** [comment]
you could do thsame on your big Sent
I think the mac mail counts are wonky

**Comment by johnstonesnow** [comment]
Yes it wouldn't surprise me to find the error was on the mac side

bash: out.txt: command not found...

**Comment by johnstonesnow** [comment]
Please don't overdo it on this

**Comment by johnstonesnow** [comment]
Well the re-index of Mac Mail failed. Gonna have to work on clearing some room, this SSD is so full. Will work on that. 
Interestingly, when I checked trash (first place i check to clear space on SSD) I see mbox files in there, and some funny names/appendices to folders. Not honestly sure why there is any of these in trash at all, I don't remember deleting any. Wondered if it's part of the conversion script's process? Just curious.

Here's a screenshot of a few of them in trash, no idea where those numbers on folder names came from!

![Screenshot 2024-03-08 at 17 35 53](https://github.com/thunderbird/import-export-tools-ng/assets/44774538/b1c9b495-f959-49a8-a79c-18a22debbdc3)

**Comment by cleidigh** [comment]
Afternoon...
Nope ietng wont create anything.
The From count from grep is clear, the Mac mail counts are not matching what the export is. I would be interested in the grep on your senf mbox.
Anyway, what would you like to do further?

**Comment by johnstonesnow** [comment]
Hi Christopher. 
well I'd like to get my emails from Mac to Thunderbird on Linux :) 
From what you said there, am I right in thinking you believe the export/conversion/import happened properly, it's just the Mac Mail reporting incorrect numbers of msgs in folders? If so, I guess I can confirm that by manually counting them. If that's the case I guess I'm good to go, if I can get some confidence in that it would be good!
Did you say there's a new version of the script? I am wondering about doing this, to try to gain confidence that I have all the emails:

1. Re-index Mac Mail - if folder counts are off, that is the fix so should solve it.
2. Reconvert it with your delightful script :)
3. Clear down TB and reimport from new conversion

What are your thoughts on this please?

**Comment by cleidigh** [comment]
I would definitely do the rebuild and reindex. Then do the import again and we can see if there are any problems after that.
Christopher

**Comment by johnstonesnow** [comment]
Roger that, thanks

**Comment by johnstonesnow** [comment]
Just a thought... Is there any chance that grep count command would work on Mac? If so, I could try counting the Mac Mail exported folders just to see what numbers come back in terminal

**Comment by johnstonesnow** [comment]
I missed your comment about being interested in grep count for Sent folder. Here's what I reported earlier for Mac v TB:

Mac - 37,374 msgs
TB - 5116 msgs

Here's what that grep count command returned for Sent folder: 5176

Looking like you're right about wonky Mac counts!

**Comment by johnstonesnow** [comment]
PS not sure if its relevant but Mac Mail claims to be using 36.33GB of space. The exptree folder i have on Fedora contains 9.8GB of data according to Files app (right click properties on folder). 

Just in case that is indicative of a failure somewhere, unless the files in exptree is compressed

**Comment by johnstonesnow** [comment]
I just looked at folder properties on two folders:

1. The folder on Mac where i put the entire Mac Mail export, before conversion. 
2. The exptree folder on Linux which I brought over after conversion

9.81GB (Mac) and 9.8GB (Fedora) - suggests all data came over, so any failure would have to be Mac Mail's fault i think.

**Comment by johnstonesnow** [comment]
**WELL would you believe it!**

Firstly, the most incredible part: I think I managed to 'hack' my way through an exercise which could be VERY insightful here. You'll laugh when you see how simple what I 'achieved' was (if I did indeed achieve what I think I did!)

Secondly... I wanted to try doing the grep/count command on Mac, on matching folders from the mac export (before script run and converted). 

I changed directory to do what we did on LInux, but soon learned (when  Mac terminal reported it) that the .mbox files are actually directories. So the same command won't work, but I then went a level deeper into the .mbox folder, to see two files in each (mbox, table of contents). 

I thought a count on the mbox would be the same as the counts we did on linux, though I confess I was just guessing my way through hoping my logic was't too far off. 

I ran grep -c 'From ' 'mbox' once inside each of the three folders I counted yesterday on Fedora. 

Well lookie here:

Trash Archive - 9948
Sent Archive - 5176
Keep Archive - 1624

The numbers I got doing the grep on Linux (on converted files in exptree folder) were:

Trash - 9948
Sent - 5176
Keep - 1627 (3 more on Linux there)

Have a I made as much progress as I think? (In confirming your suspicion of wonky Mac Mail msg counts?!

**Comment by cleidigh** [comment]
Yes, I think that is very clear here, good sleuthing.
Don't forget any small deltas are message From... Lines in bodies.
One thing would be to go to one of the folders in MM and see what the count is when you select all.
Regardless, The mbox files exported are being imported with whatever is in them.
Christopher

**Comment by cleidigh** [comment]
No not OCD. The massive delta in the counts presumably is because of bad housekeeping in MM. Something like not reseting a counter after a big delete. If you really need more you could check the actual mbox files used in the system as opposed to the export.
Christopher

**Comment by johnstonesnow** [comment]
I like your thinking! I'll have a go at that.
Overnight I will run a reindex, and tomorrow I expect to see better numbers but will check live mbox files too. One way or another I think I will suss it now, will let you know the outcome, and I'll be visiting the pluginpage as mentioned earlier, over next few days when I get a minute free to express my gratitude. 
Take care for now Christoper and thanks again.

**Comment by cleidigh** [comment]
sounds good, but hopefully you can do something better with your weekend  %-}

**Comment by johnstonesnow** [comment]
Yes, remember what I said about OCD :-D

By the way, I just found the folder containing all mail data for Mac Mail. It claims to contain 31.67GB of data. That's worrying. 

I guess it's possible that Mac Mail screwed up its own export, meanwhile your script converted everything it was asked to, and that's where the discrepancy lies.

Reindex first. Then export again. I may also try taking the live maildata folder to the better iMac, create a temp user to push that data into, then try exporting from mac mail there, would help me rule out this old dog of a Mac faiing on export. 

_"If you really need more you could check the actual mbox files used in the system as opposed to the export."_ - I just tried that. when i go to the directory in terminal, even just 'ls' command fails with "Operation not permitted". Maybe it's a 'mac thing' that you can't dig in that much. MM is not running so can't see why I am not allowed. But more to the point, I looked in Finder and of course prior to export it's all emlx files in a huge nested tree of folders. 

Will just go with a reindex and hope things become clearer. Thanks again C

PS Ha, what's a weekend?!

**Comment by cleidigh** [comment]
far, far away in an unknown galaxy and planet, there was a place where people actually valued there time and actually named a part of their week, just for pleasure and "chilling"

...Never heard of it.

**Comment by johnstonesnow** [comment]
hahaha. you and me both.

**Comment by johnstonesnow** [comment]
Well the reindex wouldn't work but I eventually found out it works better in Safe Mode. Did that and successfully reindexed all Mail data on Mac. Then I ran a new full export of all local folders, just as last time. But last time the total size of export was 9.81GB. This time it's 10.27, that's a fair bit more, and i haven't added any emails to those folders, so maybe the reindex helped. Soon find out, going to try the conversion script again overnight and see what it looks like when i import the new set to Thunderbird. Will post the result

**Comment by cleidigh** [comment]
That delta is odd.
Will be interesting to see exports.
Christopher

**Comment by johnstonesnow** [comment]
It's all very odd! Here's the latest...

I did as i said above, did a fresh python conversion, and a fresh import into a new folder created for the purpose in TB. 
This time, again all the smaller folders have identical numbers to MM. But the few big ones we've had trouble with, whilst STILL HAVING LESS than MM, had several thousand more this time!

Example, Sent Items Archive:
MM - 8861
TB after our first attempt said 5116
This time round: 8855

So this second batch seems to have fixed that folder and got them all this time. 

I have a Trash Archive I am still struggling with. 13925 on MM. 9672 on TB first time, exactly the same second time. 
I am now running a new import JUST of that one folder, to see if more  'make it' into TB. Assuming they are in the file, I think they are. (Edit: This still only brought in 9673, not 13k+ but I wonder if it may just be TB bugging out a bit, for example between conversations/msgs. Either that or the script doesn't like folders with more than 10k msgs  inside.

So currently I am wondering if it's just too much to ask of the script, these huge folders with over 10k msgs inside. 

Truth is, after refreshing my head and taking a renewed look at the situation, all the most important stuff looks to be in TB. Those old archive folders are just there for future legal reasons, so they are not unimportant, but I don't actually NEED them on my fedora laptop. So I am thinking worst case I just do this:

Store MM export AND the converted exptree on a USB stick in firesafe. I can import (either to a mac or TB) in the unilkely event I need to trawl back through those old mails any time in future.

**Comment by johnstonesnow** [comment]
PS you probably would know the answer to a related curiosity question...

For long term storage as mentioned above, what would you say is the most 'stable' medium, USB stick or old Toshiba HDD?!

**Comment by johnstonesnow** [comment]
Ok, I think I have everything sorted out. 

Won't bore you with the details but I did lots of manual checks and can't find any missing emails, it's possible there are, but I took backups to put in the firesafe (of MM and TB) so I do 'have' it all if ever need to find a very old email in the future. 

I used duplicate remover addon to clean up and it did find a good few thousand dupes. I still suspect Mac email counts could have been 'wonky' as you first suspected. Either way, I have everything secured (if not on this machine) and so I now have a good working Thunderbird installation with all my important emails in the folders they should be in :-)

Thank you again Christopher, you were beyond superb here.

**Comment by cleidigh** [comment]
Mark
Thank you so much for your kind contribution. However, your review is really incredibly heartfelt. As I indicated before it's simply a pleasure working with someone who is patient and understanding that these types of issues have multiple variables that may or may not involve IETNG or Thunderbird.

I have enjoyed working together on this and of course am here if you run into anything else.

Follow up: 

For storage, probably  archival quality optical discs then high quality HDD. Definitely not solid state devices.

I can get you the more complex grep regex that distinguishes between message From separators and From in bodies. When you get counts like you do it should be obvious.

On the script, it's actually not written by me, but @obar who really deserves a lot of credit (thanks @obar!)
It's also important to understand the scipt does not modify the mbox files, it just puts them in a different directory.

Neither IETNG or Thunderbird have issues with file sizes as yours. I have users with 7-12GBs

I think all but the Trash is on target and I just think it's a mac mail thing. We proved whatever is being exported is being imported.

Finally, isn't it also possible to export to individual eml files? That would allow another comparison.
Christopher

**Comment by johnstonesnow** [comment]
Appreciate your comments fully Christoper, but as a recipient of such great help (rare in my experience for 'noobs' like me), and with you being the giver of such help, I argue strongly that I am the most pleased beneficiary of the interaction. :P 

"Definitely not solid state devices." - Wow, glad I asked! I'd have thought they were the most 'stable' due to least moving parts, but your reply shows why I felt it best to check my 'assumptions', those being the mother of all mistakes of course!

Thank you to @obar from me too!

Re eml files - when I get some time I may have a look into that. For now, I feel I can finally have a bit of confidence in this machine and can finally STOP using the imac I have long wanted to use for target practice :). (Backups will be in safe in both folder structures so i should be good i think regardless of what weirdness was going on!)

**Comment by cleidigh** [comment]
Excellent.
The reasons against SSDs are that they use NAND gates which have a finite write/erase cycles and can have breakdown over extended times. HDDs do have mechanical aspects, however, in a non powered scenario, the media approaches tape drives, the gold standard.
C

**Comment by johnstonesnow** [comment]
Thanks!
I see, very interesting indeed.
25-30 years ago I worked in a datacentre loading large magnetic tapes all day and night (longest shift was 36 hrs straight, nearly rolled car on way home, never did a 36 hr stint again!) 
So a decent quality HDD is about as good as I can get with what I have here (and can't buy gear just for this). I will format a hard drive, look for a way to encrypt the file, and get it stored. thanks!

**Comment by johnstonesnow** [comment]
Yes, it did take me a minute or so. I'm English after all.  :-P
